pam/integration-tests/gdm-module-handler_test.go

```diff
@@ -0,0 +1,236 @@
+package main_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"slices"
+	"testing"
+
+	"github.com/msteinert/pam/v2"
+	"github.com/stretchr/testify/require"
+	"github.com/ubuntu/authd"
+	"github.com/ubuntu/authd/internal/log"
+	"github.com/ubuntu/authd/pam/internal/gdm"
+	"github.com/ubuntu/authd/pam/internal/gdm_test"
+	"github.com/ubuntu/authd/pam/internal/proto"
+)
+
+type gdmTestModuleHandler struct {
+	t  *testing.T
+	tx *pam.Transaction
+
+	protoVersion uint32
+
+	supportedLayouts []*authd.UILayout
+
+	currentStage  proto.Stage
+	pollResponses []*gdm.EventData
+
+	authModes           []*authd.GAMResponse_AuthenticationMode
+	authModeID          string
+	selectedAuthModeIDs []string
+
+	brokersInfos       []*authd.ABResponse_BrokerInfo
+	brokerID           string
+	selectedBrokerName string
+
+	eventPollResponses map[gdm.EventType][]*gdm.EventData
+
+	pamInfoMessages  []string
+	pamErrorMessages []string
+}
+
+func (gh *gdmTestModuleHandler) exampleHandleGdmData(gdmData *gdm.Data) (*gdm.Data, error) {
+	switch gdmData.Type {
+	case gdm.DataType_hello:
+		return &gdm.Data{
+			Type:  gdm.DataType_hello,
+			Hello: &gdm.HelloData{Version: gh.protoVersion},
+		}, nil
+
+	case gdm.DataType_request:
+		return gh.exampleHandleAuthDRequest(gdmData)
+
+	case gdm.DataType_poll:
+		responses := gh.pollResponses
+		gh.pollResponses = nil
+		return &gdm.Data{
+			Type:         gdm.DataType_pollResponse,
+			PollResponse: responses,
+		}, nil
+
+	case gdm.DataType_event:
+		err := gh.exampleHandleEvent(gdmData.Event)
+		if err != nil {
+			return nil, err
+		}
+		return &gdm.Data{
+			Type: gdm.DataType_eventAck,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("unhandled protocol message %s",
+		gdmData.Type.String())
+}
+
+func (gh *gdmTestModuleHandler) exampleHandleEvent(event *gdm.EventData) error {
+	events, ok := gh.eventPollResponses[event.Type]
+	if ok && len(events) > 0 {
+		pollResp := events[0]
+		gh.eventPollResponses[event.Type] = slices.Delete(events, 0, 1)
+		gh.pollResponses = append(gh.pollResponses, pollResp)
+	}
+
+	switch ev := event.Data.(type) {
+	case *gdm.EventData_BrokersReceived:
+		if len(ev.BrokersReceived.BrokersInfos) == 0 {
+			return errors.New("no brokers available")
+		}
+		gh.brokersInfos = ev.BrokersReceived.BrokersInfos
+
+		if gh.selectedBrokerName == ignoredBrokerName {
+			return nil
+		}
+
+		idx := slices.IndexFunc(gh.brokersInfos, func(bi *authd.ABResponse_BrokerInfo) bool {
```

Comment:
you can use `slices.ContainsFunc`

Reply:
Eh, I do use the index to access to the broker once there...

Reply:
sorry, I stopped as < 0, you are correct.